### PR TITLE
Automated reporting

### PR DESCRIPTION
**Pull request by krf**
Allow to print reports from command-line
    
    This can be used to quickly create reports. In combination with being
    able to quickly change the saved searches using the command-line one can
    automatically create reports e.g. after each work day.
    
    Example invocation:
      fatcrm --config-key savedSearch-1/modifiedBefore --set-config-value 2019-07-02
      fatcrm --config-key savedSearch-1/modifiedAfter --set-config-value 2019-07-02
      fatcrm --load-saved-search "Opportunities changed today" --print-to-pdf ~/out.pdf
    
     (first modify the saved search, then load and print it)

### CODE REVIEW

File: client/src/app/mainwindow.cpp [hunk 1]
```diff
@@ -605,8 +611,18 @@ void MainWindow::slotConfigure()
...
+    if (!page) {
+        return;
+    }
+
+    auto report = page->generateReport();
+
+    KDReports::PreviewDialog preview(report.get(), this);
+    preview.setWindowTitle(i18n("Print Preview"));
+    preview.previewWidget()->setShowPageListWidget(false);
+    preview.previewWidget()->setShowTableSettingsDialog(false);
```
Comment (dfaure-kdab) [inline client/src/app/mainwindow.cpp]: (pre-existing) this line is a no-op, fatcrm doesn't use the mode KDReports::Report::SpreadSheet.

File: client/src/utilities/enums.cpp [hunk 2]
```diff
@@ -23,6 +23,7 @@
 #include "enums.h"
 #include <KLocalizedString>
 #include <QDebug>
+#include <QMetaEnum>
```
Comment (dfaure-kdab) [inline client/src/utilities/enums.cpp]: unused?

File: client/src/reports/reportgenerator.cpp [hunk 3]
```diff
@@ -73,34 +76,30 @@ void ReportGenerator::addHeader(KDReports::Report &report)
...
+    report->addElement(table);
+
+    finalizeReport(*report);
 
-    finalizeReport(report, parent);
+    return report;
 }
 
-void ReportGenerator::finalizeReport(KDReports::Report &report, QWidget *parent)
+void ReportGenerator::finalizeReport(KDReports::Report &report)
```
Comment (dfaure-kdab) [inline client/src/reports/reportgenerator.cpp]: this method isn't really worth it anymore, for a single line of code, called from a single location...

File: client/src/app/main.cpp [hunk 4]
```diff
@@ -55,6 +56,10 @@ int main(int argc, char **argv)
     KAboutData::setApplicationData(aboutData);
     parser.addVersionOption();
     parser.addHelpOption();
+    QCommandLineOption printToPdfOption("print-to-pdf", i18n("Print page after initial import and exit FatCRM"), "fileName");
+    parser.addOption(printToPdfOption);
+    QCommandLineOption printTypeOption("print-type", i18n("Print type, either 'Accounts', 'Opportunities' or 'Contacts', ..."), "printType", "Opportunities");
```
Comment (dfaure-kdab) [inline client/src/app/main.cpp]: Remove "or" given the "..." at the end? The grammar seems odd to me.

**Comment by krf** [comment]
> "quickly change the saved searches using the command-line" -- is that implemented?

Yep, I've updated the PR message to show some more details how to use this feature.

File: client/src/app/main.cpp [hunk 5]
```diff
@@ -53,10 +57,40 @@ int main(int argc, char **argv)
     parser.addHelpOption();
     QCommandLineOption noOverlayOption("nooverlay", i18n("Do not display the overlay during initial data loading"));
     parser.addOption(noOverlayOption);
+    QCommandLineOption configKeyOption("config-key", i18n("Internal feature: Return the value for setting <key>"), "key");
+    parser.addOption(configKeyOption);
+    QCommandLineOption configValueOption("config-value", i18n("Internal feature: Set the value for setting specified by 'configKey' parameter"), "value");
```
Comment (dfaure-kdab) [inline client/src/app/main.cpp]: Ah, how about naming this option "set-config-value" ? It wasn't obvious when reading your example command line that this was a setter, not a getter.

**Comment by krf** [comment]
Okay to merge?

**Pull request merged**